packages/slate/src/interfaces/custom-extensions.ts

```diff
@@ -0,0 +1,13 @@
+/**
+ * The `CustomExtensions` interface extends types
+ */
+
+export interface CustomExtensions {
```

Comment:
I still think we need to update Slate's types for `Text` and `Element` to use `ExtendedType` below?  See @thesunny suggestion in the implementation issue - https://github.com/ianstormtaylor/slate/issues/3725.  For instance:

```
// This would be Element as per Slate's definition
export type BaseElement = {
  children: Text[]
}

// This would be Text as per Slate's definition
export type BaseText = {
  text: string
}

export type Text = ExtendedType<"Text", BaseText>
export type Element = ExtendedType<"Element", BaseElement>
```

Then, the exported new types for `Text` and `Element` would be used in Slate's codebase and any developer's use of `CustomExtensions` and `ExtendedType` would propagate to the rest of the codebase.  

Let's discuss if needed.

Reply:
The issue with this is that when you use `ExtendedType` you are redefining Element or Text according to the definition you gave it inside your CustomExtension. If make the extended Element the new Element, a bunch of errors will be thrown during build, and pretty much 60% of the testcases will fail. Something that I found while making this PR was that in order to call Element's functions like `isElement,` you need to use an alias. If you take a look at my isElement test case, you will see I am calling BaseElement.isElement. If I were to use Element.isElement, the program would not recognize the function because the type Extension is not able to fetch the functions as well.

Reply:
Sorry, I haven't had time to dive into the code so perhaps you could explain the issue and I might be able to help with a solution.

My question is, if the original object is defined as BaseText or BaseElement, and the extended type is created as Text and Element, are you saying that the methods don't come through to the extend type? It looks like in the typing we are doing:

```ts
export type ExtendedType<K extends string, B> =
  unknown extends CustomExtensions[K]
  ? B
  : B & CustomExtensions[K]
```

So shouldn't we get the proper types from both the original with the extensions added (ie. `B & CustomExtensions[K]`)?

I may be misunderstanding the issue in which case a clarification would be helpful. Thanks.

Reply:
@thesunny could use a review from your end.  I think the code is looking more final now so it's probably a good time to review.  Thanks!

@CameronAckermanSEL can you review too?